src/qasm-simulator-cpp/src/utilities/tensor_index.hpp

```diff
@@ -100,29 +93,32 @@ TensorIndex::TensorIndex() {
     bits[i] = (1ULL << i);
 }
 
+
+
 //------------------------------------------------------------------------------
 // Static Indexing
 //------------------------------------------------------------------------------
 
 template <size_t N>
 uint_t TensorIndex::index0_static(const std::array<uint_t, N> &qubits_sorted,
                                  const uint_t k) const {
-  uint_t lowbits = 0, mask = 0;
+  uint_t lowbits, retval = k;
   for (size_t j = 0; j < N; j++) {
-    mask ^= masks[qubits_sorted[j] - j];
-    lowbits |= (k & mask) << j;
+    lowbits = retval & masks[qubits_sorted[j]];
+    retval >>= qubits_sorted[j];
+    retval <<= qubits_sorted[j] + 1;
```

Comment:
I can't see the difference between these two shifts:
``` 
    retval >>= qubits_sorted[j];
    retval <<= qubits_sorted[j] + 1;
```
and just:
```
retval <<= 1;
```
Am I wrong?

Reply:
It's necessary. We are taking an arbitrary bit-string and inserting a 0 into a specified location by:
* shifting down to clear lower bits (`retval >>= qubits_sorted[j]`)
* shifting up to insert a bit leaving it and all lowbits in zero (`retval <<= qubits_sorted[j] + 1`). 
* Then the lowbits are reset to initial value (`retval |= lowbits`)
Just shifting would insert the zero bit at the start shifting all bits over 1.

Reply:
wow, ok!